### PR TITLE
Fix unsafe access to a vector in cuSTL MPI reduce

### PR DESCRIPTION
**Pull request by sbastrakov**
Variable `new_ranks` can be empty, and so `new_ranks.front()` is incorrect without a bounds check. This situation occurs, for example, if there is a single process and `setThisAsRoot` is false.
So add a bounds check before the unsafe code.

This was found during investigation of #3329, as it crashed on my machine, and is an actual bug that this PR fixes. But this is not what caused it on Hemera.

### CODE REVIEW

File: include/pmacc/cuSTL/algorithm/mpi/Reduce.tpp [hunk 1]
```diff
@@ -80,11 +80,13 @@ Reduce<dim>::Reduce(const zone::SphericZone<dim>& p_zone, bool setThisAsRoot) :
...
     // avoid deadlock between not finished pmacc tasks and mpi blocking collectives
     __getTransactionEvent().waitForFinished();
-    MPI_CHECK(MPI_Comm_group(MPI_COMM_WORLD, &world_group));
-    MPI_CHECK(MPI_Group_incl(world_group, new_ranks.size(), &(new_ranks.front()), &new_group));
-    MPI_CHECK(MPI_Comm_create(MPI_COMM_WORLD, new_group, &this->comm));
-    MPI_CHECK(MPI_Group_free(&new_group));
-    MPI_CHECK(MPI_Group_free(&world_group));
+    if(new_ranks.size())
+    {
+        MPI_CHECK(MPI_Group_incl(world_group, new_ranks.size(), &(new_ranks.front()), &new_group));
```
Comment (n01r) [inline include/pmacc/cuSTL/algorithm/mpi/Reduce.tpp:85]: Why did you delete `MPI_CHECK(MPI_Comm_group(MPI_COMM_WORLD, &world_group));` in the line before?

Comment (sbastrakov) [reply include/pmacc/cuSTL/algorithm/mpi/Reduce.tpp:85]: Ah, good catch. My mistake, it should stay in place.

Comment (sbastrakov) [reply include/pmacc/cuSTL/algorithm/mpi/Reduce.tpp:85]: Fixed and rebased. Also moved declarations of those two variables inside the `if` statement as they are only used there.

Comment (n01r) [reply include/pmacc/cuSTL/algorithm/mpi/Reduce.tpp:85]: That would have been my next question ^^

**Review by n01r — APPROVED** [review]
Alright, looks good :)

**Comment by sbastrakov** [comment]
@n01r could you press the button?
